### DISCUSSION TITLE
Usage of ffmpeg hls encoder

### DISCUSSION THREAD

**Discussion by mp3butcher** [savonet/liquidsoap#3669] [General]
Is there any reason why hls is implemented natively and not with the  ffmpeg hls encoder?

**Comment by toots** [comment]
Yes, the native implementation gives us much more flexibility w.r.t. to for instance the segment naming, using callbacks when segments have finished being prepared and etc.

You should still be able to use `ffmpeg`s own implementation using the proper options in your `%ffmpeg` output.

**Comment by mp3butcher** [comment]
i tried to reproduce this simple command line
`ffmpeg -re  -i track.mp3 -c:a aac -map 0:0 -f hls  -hls_list_size 10   -master_pl_name "livemaster.m3u8"  live.m3u8`
i tested in liquidsoap 2.0:
`enc = %ffmpeg(format="hls", %audio( codec="aac", channels=2, ar=44100, b="128k"), hls_list_size=10, "master_pl_name"="livemaster.m3u8")`

butI can't figure how to get ffmpeg output fragment with   `output.file( enc,"{{ paths.hls_output_path }}/main.m3u8", s)`

**Comment by toots** [comment]
This works:
```liquidsoap
enc =
  %ffmpeg(
    format = "hls",
    %audio(codec = "aac", channels = 2, ar = 44100, b = "128k"),
    hls_list_size = 10,
    master_pl_name = "livemaster.m3u8"
  )

output.url(url="/tmp/hls/radio.m3u8", enc, input, fallible=true)
```

**Reply by mp3butcher** [reply]
Your code works, I can play it in vlc but not in a conventional HTML5 player.
hls requires multiple files to be output and accessible by the player via a web server.
Output.file and output.url  seems to target only the ffmpeg output file. 
Perhaps my proposal to add an output/file/ffmpeg is not so dumb.

**Reply by toots** [reply]
I think what you need is to point your webserver to the directory where those files are created.

With our internal output, and the reason we're doing it, you can use the callbacks to instead upload the files to a cdn typically.

HLS is a complex protocol that mixes encoding, format and transport considerations. Ffmpeg is really good at all these things in isolation but I'm not sure it's the most appropriate tool to work with the weird combination of these that HLS requires. 

Typically ffmpeg will create segments that are over the segment target duration, something that the specs prohibit.

**Reply by mp3butcher** [reply]
> I think what you need is to point your webserver to the directory where those files are created.

Already done...that's why i say it works (curiously) in vlc as client but not in web player.. 
I'm quite familiar with hls protocol (I worked a lot on youtube_dl) but not with ocaml.
it requires a main manifest several children manifest and fragments to be accessed through http by client.
Your solution with output.url  may outputs concatenated data in http socket (that might be why only vlc achieve to read this) but it doesn't produce neither of the required files.
 
> Typically ffmpeg will create segments that are over the segment target duration, something that the specs prohibit.

it might be  fixed https://trac.ffmpeg.org/ticket/8037 at least with latest ffmpeg release https://trac.ffmpeg.org/ticket/9321

**Reply by toots** [reply]
That's not what I'm seeing. When I run the code above I see the main playlist, stream playlist and segments being created at the destination.

What you are experiencing is most likely due to the fact that chrome does not have native support for HLS: https://caniuse.com/http-live-streaming

You may need to use [hls.js](https://github.com/video-dev/hls.js) to enable HLS support on your `<audio>` element.

You can also try with safari which should have native support for it.

Happy to give more details, I am currently working on something like that for video.

**Comment by mp3butcher** [comment]
@toots  Mea Culpa, Your solution works as it should. I tested it too quickly. 
Thanks for your support
